docs/kb/edk2-capsule-updates.md

```diff
@@ -325,6 +331,331 @@ interferes).
 
 [wiki-pkc]: https://en.wikipedia.org/wiki/Public-key_cryptography
 
+## Generating signing keys with OpenSSL
+
+The process involves creation of 3 certificates, a local certificate
+authority (CA) in a directory and 14 files.  This warrants some overview of
+what's going to be done.
+
+Keys making up the chain will assume the following roles:
+
+- _root_ — self-signed certificate acting as the basis of a CA
+- _sub_ — second-level CA signed by _root_ key
+- _sign_ — a signing key (not a CA) that is signed by _sub_ key
+
+A digital signature algorithm (DSA) consists of a private key (PK) algorithm and
+a digest algorithm.  Keys in a chain don't have to use the same DSA.  Also, most
+DSAs allow use of an arbitrary digest (not Ed25519 or Ed448 which prescribe the
+use of SHA-512 and SHAKE-256 respectively).
+
+Keys exist independently of the chains in which they appear.  When stored,
+private keys are often encrypted.
+
+### Key generation by example
+
+It will be easier to go into finer details while looking at the actual commands.
+The entire process can be broken into several stages:
+
+1. Creation of 3 certificates
+2. Construction of a CA
+3. Preparation of _root_ certificate for inclusion into EDK
+4. Preparation of _sign_ certificate for signing
+
+#### Make certificates
+
+Using 4096-bit RSA keys as an example:
+
+```bash
+openssl genrsa -aes256 -out root.p8e 4096
+openssl genrsa -aes256 -out sub.p8e 4096
+openssl genrsa -aes256 -out sign.p8e 4096
+```
+
+Each of them is encrypted with AES-256.  The password is queried interactively
+for creation and each time the certificate is used to sign something.  Drop
+`-aes256` to not use encryption (for a test or if you consider access to the
+files a complete compromise of the security).
+
+`.p8e` extension is for PKCS #8 format carrying an encrypted private key.
+
+#### Make a CA
+
+By default, directory for a CA is called `demoCA` (can be changed in
```

Comment:
In my case the default directory was `/etc/ssl/pki/CA`. I had to either change it in the openssl.conf or move the contents of `demoCA` to that directory.

Reply:
I can confirm, this is a difference between Fedora and coreboot-sdk:

```
[khebel@3M05 coreboot]$ grep "^dir" /etc/ssl/openssl.cnf
dir		= /etc/pki/CA		# Where everything is kept
dir		= /etc/pki/CA		# TSA root directory
[khebel@3M05 coreboot]$ docker run --rm -it -u $UID -v $PWD:/home/coreboot/coreboot     -v $HOME/.ssh:/home/coreboot/.ssh     -w /home/coreboot/coreboot coreboot/coreboot-sdk:2023-11-24_2731fa619b     /bin/bash
coreboot@edfefef97a56:~/coreboot$ grep "^dir" /etc/ssl/openssl.cnf
dir		= ./demoCA		# Where everything is kept
dir		= ./demoCA		# TSA root directory
```

Interestingly, `man openssl-sa` on Fedora still mentions `demoCA`:

```
FILES
       Note: the location of all files  can  change  either  by  compile  time
       options,  configuration  file entries, environment variables or command
       line options.  The values below reflect the default values.

        /usr/local/ssl/lib/openssl.cnf - master configuration file
        ./demoCA                       - main CA directory
        ./demoCA/cacert.pem            - CA certificate
        ./demoCA/private/cakey.pem     - CA private key
        ./demoCA/serial                - CA serial number file
        ./demoCA/serial.old            - CA serial number backup file
        ./demoCA/index.txt             - CA text database file
        ./demoCA/index.txt.old         - CA text database backup file
        ./demoCA/certs                 - certificate output file
```

Reply:
I have updated `capsule.sh` to set its own directory name so it should work everywhere, but it's not exactly user-friendly, so here I just mentioned that it can be different and where to look it up.

https://github.com/Dasharo/docs/compare/f97a4be31c9c8f3a5f80f0f4d4542055d128e92b..ccd16f85cca0082c07da3e86bc4e32999603d3d7#diff-b0d92435e1999e660dd9b88bd5bd5dd14331f32b3abce08c0b9a5895537af576R383-R387

Reply:
`capsule.sh` does work fine using its own directory. Generating the keys and the CA manually on Fedora still would not work when following the instructions directly and would require adjusting the CA Default directory or using the one already present in openssl.conf but I guess it's fine to just tell that there may be differences. We don't need to maintain a guide suited for every distribution.